File: Data Structures & Algorithms/three-integer-sum/submission-5.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        i = 0 
        res = []
        while i < len(nums):
            if nums[i] >= 0:
                break
            l = i + 1
            r = len(nums) - 1
            while l < r:
                sum = nums[i] + nums[l] + nums[r]
                if sum == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while l < len(nums) and nums[l] == nums[l - 1]:
                        l += 1
                elif sum > 0:
                    r -= 1
                else :
                    l += 1
            i += 1
            while i < len(nums) and nums[i] == nums[i - 1]:
                i+=1
        return res

File: Data Structures & Algorithms/combination-target-sum-ii/submission-5.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        res.clear();
        sort(candidates.begin(), candidates.end());
        dfs(0, {}, 0, candidates, target);
        return res;
    }

private:
    void dfs(int idx, vector<int> path, int cur, vector<int>& nums, int target) {
        if (cur == target) {
            res.push_back(path);
            return;
        }
        if (cur > target || idx >= nums.size()) return;
        
        path.push_back(nums[idx]);
        dfs(idx + 1, path, cur + nums[idx], nums, target);
        path.pop_back();
        idx++;
        while (idx < nums.size() && nums[idx] == nums[idx - 1]) idx++;
        dfs(idx, path, cur, nums, target);

    }
};